### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by ghost**
👋

Should probably use benchee but

Current implementation
```elixir
iex(1)> :timer.tc fn -> EmailGuard.check("aksdjfhg@gmail.com") end
{122, :ok}
iex(2)> :timer.tc fn -> EmailGuard.check("aksdjfhg@gmail.com") end
{68, :ok}
iex(3)> :timer.tc fn -> EmailGuard.check("aksdjfhg@gmail.com") end
{69, :ok}
```

This PR
```elixir
iex(1)> :timer.tc fn -> EmailGuard.check("aksdjfhg@gmail.com") end
{13, :ok}
iex(2)> :timer.tc fn -> EmailGuard.check("aksdjfhg@gmail.com") end
{12, :ok}
iex(3)> :timer.tc fn -> EmailGuard.check("aksdjfhg@gmail.com") end
{14, :ok}
```

Can probably be optimized further ... I had a library of "bad" words, and lookups there took ~4us.

---

```elixir
  "../../priv/disposable_list.txt"
  |> File.stream!()
  |> Stream.map(&String.trim/1)
  |> Stream.filter(&match?("", &1))
  |> Enum.map(fn domain ->
    def lookup(unquote(domain)), do: "DisposableList"
  end)
```
for each list can be replaced with a macro.

```elixir
defmodule EmailGuard.DisposableList do
  use List, list_path: "../../priv/..."
end
```

So that we would need to provide just the path to the file with "bad" domains.

### CODE REVIEW

**Comment by ghost** [comment]
Oopsie. It doesn't actually work.

**Comment by svileng** [comment]
Awesome, that's pretty nice 😎 

Could you please squash the commits into 1 and give the final commit a name to reflect the PR, e.g. "Performance improvements" or similar? I will be happy to merge it afterwards and prepare a release 👍 

Thanks a lot for spending time on this 🍻

**Comment by ghost** [comment]
> Could you please squash the commits into 1 and give the final commit a name to reflect the PR, e.g. "Performance improvements" or similar?

@svileng sure, but what about a `use` macro? Would you want it in this PR or another?

I would also like to use benchee first to be sure there is a significant improvement.

**Comment by svileng** [comment]
You can include it in this PR if you have time, I certainly don't mind. If not then I can add it afterwards. Whichever is easier 👍

**Comment by ghost** [comment]
@svileng benchee results

```
Operating System: macOS"
CPU Information: Intel(R) Core(TM) i5-4258U CPU @ 2.40GHz
Number of Available Cores: 4
Available memory: 8 GB
Elixir 1.6.4
Erlang 20.3.2

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 56 s


Benchmarking disposable found...
Benchmarking disposable passed...
Benchmarking free found...
Benchmarking free passed...

Name                        ips        average  deviation         median         99th %
free passed            979.93 K        1.02 μs  ±5762.95%           1 μs           2 μs
disposable passed      972.29 K        1.03 μs  ±5684.23%           1 μs           2 μs
free found             970.02 K        1.03 μs  ±5184.01%           1 μs           2 μs
disposable found       832.54 K        1.20 μs  ±3962.38%           1 μs           2 μs

Comparison:
free passed            979.93 K
disposable passed      972.29 K - 1.01x slower
free found             970.02 K - 1.01x slower
disposable found       832.54 K - 1.18x slower

Memory usage statistics:

Name                 Memory usage
free passed                 832 B
disposable passed           832 B - 1.00x memory usage
free found                  904 B - 1.09x memory usage
disposable found            856 B - 1.03x memory usage

**All measurements for memory usage were the same**
```

Will test the current list implementation now.

File: README.md [hunk 1]
```diff
@@ -4,8 +4,8 @@ Elixir library for detecting disposable (burner) or personal email addresses.
 
 Comes with two lists:
 
-  - `DisposableList` largely [based on this list](https://github.com/wesbos/burner-email-providers) containing 3846 domains;
-  - `FreeList` mostly [based on this](https://gist.github.com/ammarshah/f5c2624d767f91a7cbdc4e54db8dd0bf) minus the domains already listed in `DisposableList`, total 5072 domains.
+* `DisposableList` largely [based on this list](https://github.com/wesbos/burner-email-providers) containing 3846 domains;
```
Comment (ghost) [inline README.md:7]: Oops, that was sublime, not me. Will change it back.

**Comment by ghost** [comment]
Current implementation (just added benchee to `master` and ran the same script)
```
Operating System: macOS"
CPU Information: Intel(R) Core(TM) i5-4258U CPU @ 2.40GHz
Number of Available Cores: 4
Available memory: 8 GB
Elixir 1.6.4
Erlang 20.3.2

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 56 s


Benchmarking disposable found...
Benchmarking disposable passed...
Benchmarking free found...
Benchmarking free passed...

Name                        ips        average  deviation         median         99th %
free found              63.98 K       15.63 μs   ±129.67%          15 μs          28 μs
disposable found        42.42 K       23.57 μs    ±85.60%          22 μs          42 μs
free passed             25.34 K       39.47 μs    ±30.01%          38 μs          69 μs
disposable passed       24.27 K       41.20 μs    ±44.79%          39 μs          73 μs

Comparison:
free found              63.98 K
disposable found        42.42 K - 1.51x slower
free passed             25.34 K - 2.53x slower
disposable passed       24.27 K - 2.64x slower

Memory usage statistics:

Name                 Memory usage
free found                  680 B
disposable found            688 B - 1.01x memory usage
free passed                 856 B - 1.26x memory usage
disposable passed           856 B - 1.26x memory usage

**All measurements for memory usage were the same**
```

**Comment by ghost** [comment]
@svileng can't make `def lookup(unquote(domain)), do: unquote(list_name)` return a string. It keeps returning the module name instead:

```elixir
iex(1)> EmailGuard.check("a@mailinator.com")
{:error, EmailGuard.DisposableList}
```
even though it's set to a string in source code ...

**Comment by ghost** [comment]
@svileng would you want to automatically update the domain lists by downloading them during compilation if etags change?

```sh
$ curl --head https://raw.githubusercontent.com/wesbos/burner-email-providers/master/emails.txt
...
ETag: "b3a8b020ea9b5169be68c2c7a2a8b3be402de1c1"
...
```
Or maybe add a mix task for it.

It probably can be left for another PR ...

**Comment by ghost** [comment]
While playing with `EmailGuard.check` itself, I've replaced it with
```elixir
  def check(email, [email_list]) when is_binary(email) do
    email
    |> extract_domain()
    |> email_list.lookup()
    |> case do
      nil -> :ok
      list_name -> {:error, list_name}
    end
  end

  def check(email, email_lists) when is_binary(email) and is_list(email_lists) do
    domain = extract_domain(email)

    email_lists
    |> Enum.find(fn list -> list.lookup(domain) end)
    |> case do
      nil -> :ok
      list_name -> {:error, list_name}
    end
  end
```
and got > 1m ips for some checks where just one list module is provided
```
Operating System: macOS"
CPU Information: Intel(R) Core(TM) i5-4258U CPU @ 2.40GHz
Number of Available Cores: 4
Available memory: 8 GB
Elixir 1.6.4
Erlang 20.3.2

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 56 s


Benchmarking disposable found...
Benchmarking disposable passed...
Benchmarking free found...
Benchmarking free passed...

Name                        ips        average  deviation         median         99th %
free passed              1.08 M        0.93 μs  ±6387.09%           1 μs           2 μs
free found               1.05 M        0.95 μs  ±4747.03%           1 μs           2 μs
disposable found         1.03 M        0.97 μs  ±3867.27%           1 μs           2 μs
disposable passed        0.97 M        1.03 μs  ±5791.16%           1 μs           2 μs

Comparison:
free passed              1.08 M
free found               1.05 M - 1.03x slower
disposable found         1.03 M - 1.05x slower
disposable passed        0.97 M - 1.11x slower

Memory usage statistics:

Name                 Memory usage
free passed                 776 B
free found                  848 B - 1.09x memory usage
disposable found            800 B - 1.03x memory usage
disposable passed           776 B - 1.00x memory usage

**All measurements for memory usage were the same**
```

**Comment by svileng** [comment]
> can't make def lookup(unquote(domain)), do: unquote(list_name) return a string. It keeps returning the module name instead:

That's totally fine, I'll publish it as a breaking change and as `1.1.0` 👍 

> would you want to automatically update the domain lists by downloading them during compilation if etags change?

That's a good idea, but let's leave it for a separate PR.

I'm happy with all the changes, all looks good 👍Although I have no idea what is this doing 😆 
```elixir
quote bind_quoted: [filepath: path_to_list(filename)] do
```
Need to do some more reading on macros...

Anyway, happy to merge this after everything is squashed into 1 commit 🎉

**Comment by ghost** [comment]
> Although I have no idea what is this doing 😆

`bind_quoted` makes `filepath` available inside the quoted block, I think.

> Anyway, happy to merge this after everything is squashed into 1 commit 🎉

Squashed into `perf improvements`.

**Comment by Un3qual** [comment]
Any chance using `Flow` would worth using to increase performance?

**Comment by ghost** [comment]
👋

@Un3qual can you provide an example of how it could be used in this case?

**Comment by svileng** [comment]
@idi-ot Awesome, thank you for your contribution! Published as `1.1.0` 🎉

**Comment by svileng** [comment]
@Un3qual An example would be interesting for sure, but I think the current level of performance should satisfy 99% of use cases.

**Pull request merged**